src/sql/src/kafka_util.rs

```diff
@@ -257,42 +259,60 @@ where
+                    .map_err(|_| sql_err!("kafka topic had more than {} partitions", i32::MAX))?;
 
-            let mut tpl = TopicPartitionList::with_capacity(1);
-            tpl.add_partition_range(&topic, 0, num_partitions_i32 - 1);
-            tpl.set_all_offsets(Offset::Offset(time_offset))
-                .map_err(|e| sql_err!("{}", e))?;
+                let mut tpl = TopicPartitionList::with_capacity(1);
+                tpl.add_partition_range(&topic, 0, num_partitions_i32 - 1);
+                tpl.set_all_offsets(Offset::Offset(time_offset))
+                    .map_err(|e| sql_err!("{}", e))?;
 
-            let offsets_for_times = consumer
-                .offsets_for_times(tpl, Duration::from_secs(10))
-                .map_err(|e| sql_err!("{}", e))?;
+                let offsets_for_times = consumer
+                    .offsets_for_times(tpl, Duration::from_secs(10))
+                    .map_err(|e| sql_err!("{}", e))?;
 
-            // Translate to `start_offsets`
-            let start_offsets = offsets_for_times
-                .elements()
-                .iter()
-                .map(|elem| match elem.offset() {
-                    Offset::Offset(offset) => Ok(offset),
-                    Offset::End => fetch_end_offset(&consumer, &topic, elem.partition()),
-                    _ => sql_bail!(
-                        "Unexpected offset {:?} for partition {}",
-                        elem.offset(),
-                        elem.partition()
-                    ),
-                })
-                .collect::<Result<Vec<_>, _>>()?;
+                // Translate to `start_offsets`
+                let start_offsets = offsets_for_times
+                    .elements()
+                    .iter()
+                    .map(|elem| match elem.offset() {
+                        Offset::Offset(offset) => Ok(offset),
+                        Offset::End => fetch_end_offset(&consumer, &topic, elem.partition()),
+                        _ => sql_bail!(
+                            "Unexpected offset {:?} for partition {}",
+                            elem.offset(),
+                            elem.partition()
+                        ),
+                    })
+                    .collect::<Result<Vec<_>, _>>()?;
+
+                if start_offsets.len() != num_partitions {
+                    sql_bail!(
+                        "Expected offsets for {} partitions, but received {}",
+                        num_partitions,
+                        start_offsets.len(),
+                    );
+                }
 
-            if start_offsets.len() != num_partitions {
-                sql_bail!(
-                    "Expected offsets for {} partitions, but received {}",
-                    num_partitions,
-                    start_offsets.len(),
-                );
+                return Ok(Some(start_offsets));
             }
 
-            Ok(Some(start_offsets))
+            unreachable!()
```

Comment:
This is kind of weird - took me a minute to figure out what was going on with the options above.

Maybe turn the first match into just an assert instead of unpacking, and then we can do a exhaustive match here instead of the two if-lets?

Reply:
sounds good!